### PR TITLE
Timeout should take an implicit scheduler and EC

### PR DESCRIPTION
**Pull request by rossabaker**

### CODE REVIEW

**Comment by rossabaker** [comment]
That's not the resource handling failing.  That's the timeout itself failing.  We're asserting that the 204 response got preempted by the timeout response.

**Comment by ChristopherDavenport** [comment]
I mean the error is quite long it looks like it started the delivery, and half-way through was disconnected, but returned a 204 instead of a 500 and the Blaze stage was very displeased with how it was handled. 

If that is the correct behavior I think the test will need changes to match that we now expect a 204, but why would a timeout yield a 204 rather than a 500?

**Comment by rossabaker** [comment]
Hmm.  The blaze mess is a 304, not a 204.  I don't know what's doing a 304 in our tests.  I think that might be output from a different test running concurrently?

The "timeout" is returning a 204 because it isn't timing out, I think.  If the example service returned something other than a 204, with this bug, that's what we'd see instead of the 500.  The question is why the timeout isn't winning the race.

I think this started when I simplified the way the timeout response is generated, but maybe it's just the first time this test was on that branch.  It's all a bit of a blur that I haven't had time to sort out yet.

**Comment by rossabaker** [comment]
Locally:

```
[info] Timeout Middleware should
[info]   + have no effect if the response is not delayed (73 ms)
[info]   + return a 500 error if the result takes too long (73 ms)
[info]   + return the provided response if the result takes too long (73 ms)
[info]   + handle infinite durations (4 seconds, 56 ms)
[info]   + clean up resources of the loser (280 ms)
```

Travisally:

```
[info] Timeout Middleware should
[info]   + have no effect if the response is not delayed (3 seconds, 157 ms)
[info]   + return a 500 error if the result takes too long (3 seconds, 137 ms)
[info]   + return the provided response if the result takes too long (3 seconds, 103 ms)
[info]   + handle infinite durations (6 seconds, 993 ms)
[error]   x clean up resources of the loser (3 seconds, 153 ms)
[error]    the response status '204 No Content' is not equal to '500 Internal Server Error' (TimeoutSpec.scala:54)
```

Look at those time differences!

**Comment by bfritz** [comment]
Wish I had my head around this a little better, but I think it largely boils down to Travis only has 2 cores and is more likely to block on access to disk or CPU.  Combine that with running a bunch of specs in parallel and having a million (give or take) pooled threads floating around and interacting with one another, and things get pretty bogged down.

**Comment by bfritz** [comment]
Build failure was addressed by #1126 and looks like `TimeoutSpec` passed on 2.11 and 2.12 after the increase.  Worth running another build (or restarting the old one) to get a second data point?

**Comment by ChristopherDavenport** [comment]
If that solved it, that would be nice. Definitely worth investigating, but I know better than to restart another mans build.

**Comment by rossabaker** [comment]
We're all friends here.

File: server/src/test/scala/org/http4s/server/middleware/TimeoutSpec.scala [hunk 1]
```diff
@@ -48,12 +48,7 @@ class TimeoutSpec extends Http4sSpec {
       var clean = new AtomicBoolean(false)
       val service = HttpService {
         case _ =>
-          implicit val A = Task.asyncInstance(Strategy.sequential)
-          (sleep_(200.milliseconds) ++
-            eval(NoContent()) ++
-            eval_(Task.delay(clean.set(true))))
-            .runLast
-            .map(_.getOrElse(throw new AssertionError("Should have emitted NoContent")))
+          NoContent().schedule(2.seconds) <* Task.delay(clean.set(true))
```
Comment (bfritz) [inline server/src/test/scala/org/http4s/server/middleware/TimeoutSpec.scala]: Is `<*` commonly understood?  Guessing so.

Unfortunately no scaladoc to explain it:
http://typelevel.org/cats/api/cats/syntax/CartesianOps.html#%3C*[B](fb:F[B])(implicitF:cats.Functor[F]):F[A]

To be fair, I have a much better shot at understanding the version in green than the one in red. :smile:

Comment (rossabaker) [inline server/src/test/scala/org/http4s/server/middleware/TimeoutSpec.scala]: It's, like, `*>`, but not. :trollface:  I'll put it in a for comprehension for legibility.

**Pull request merged**